File: customer_support_ai/src/data_generator.ts
/**
 * @license
 * SPDX-License-Identifier: Apache-2.0
 */

import * as fs from "fs";
import * as path from "path";
import { QueryCategory, Sentiment, SupportQuery } from "./types.js";

// Helper to get random item from list
const randomChoice = <T>(arr: T[]): T => arr[Math.floor(Math.random() * arr.length)];

// Helper to get random number in range
const randomRange = (min: number, max: number): number => {
  return Math.floor(Math.random() * (max - min + 1)) + min;
};

// Raw materials for randomization
const CITIES_WITH_STATES = [
  { city: "Austin", state: "TX" },
  { city: "Boston", state: "MA" },
  { city: "Chicago", state: "IL" },
  { city: "Denver", state: "CO" },
  { city: "El Paso", state: "TX" },
  { city: "Fort Worth", state: "TX" },
  { city: "Gary", state: "IN" },
  { city: "Houston", state: "TX" },
  { city: "Indianapolis", state: "IN" },
  { city: "Jacksonville", state: "FL" },
  { city: "Kansas City", state: "MO" },
  { city: "Los Angeles", state: "CA" },
  { city: "Miami", state: "FL" },
  { city: "Nashville", state: "TN" },
  { city: "Oakland", state: "CA" },
  { city: "Phoenix", state: "AZ" },
  { city: "Queens", state: "NY" },
  { city: "Raleigh", state: "NC" },
  { city: "Seattle", state: "WA" },
  { city: "Tampa", state: "FL" },
  { city: "Urbana", state: "IL" },
  { city: "Vancouver", state: "WA" },
  { city: "Wichita", state: "KS" },
  { city: "Xenia", state: "OH" },
  { city: "Yonkers", state: "NY" },
  { city: "Zion", state: "IL" }
];

const PRODUCTS = [
  { name: "Pro Headphones X", category: "Audio", price: 299 },
  { name: "Quantum Laptop 15", category: "Computing", price: 1499 },
  { name: "Secure Router Pro", category: "Networking", price: 199 },
  { name: "Ultra Smartwatch v2", category: "Wearables", price: 349 },
  { name: "Eco Charging Pad", category: "Accessories", price: 49 },
  { name: "Noise-Cancelling Bud S", category: "Audio", price: 129 },
  { name: "4K HDR Monitor", category: "Display", price: 599 }
];

const DEVICES = ["iPhone 14 Pro", "Samsung Galaxy S23", "MacBook Pro M2", "Dell XPS 13", "iPad Air", "Windows Desktop", "Android Tablet"];
const ERROR_CODES = ["ERR_CONNECTION_TIMED_OUT", "ERR_AUTH_FAILED_401", "ERR_DB_TIMEOUT_504", "ERR_RESOURCE_NOT_FOUND_404", "ERR_SSL_HANDSHAKE_FAILED", "ERR_INVALID_LICENSE_KEY"];
const FIRST_NAMES = ["Alex", "Jordan", "Taylor", "Morgan", "Sam", "Casey", "Jamie", "Riley", "Robin", "Drew", "Skyler", "Cameron", "Chris", "Pat", "Amelia", "Benjamin", "Chloe", "Daniel", "Emily", "Fiona", "George", "Hannah", "Ian", "Julia", "Kevin", "Lily", "Matthew", "Natalie", "Owen", "Penelope", "Quincy", "Ruby", "Samuel", "Victor", "Wyatt", "Xavier", "Yvonne", "Zachary"];
const LAST_NAMES = ["Smith", "Johnson", "Williams", "Brown", "Jones", "Miller", "Davis", "Garcia", "Rodriguez", "Wilson", "Martinez", "Anderson", "Taylor", "Adams", "Baker", "Carter", "Evans", "Fisher", "Green", "Hill", "Irvin", "King", "Lewis", "Nelson", "Ortiz", "Perez", "Quinn", "Robinson", "Thomas", "Underwood", "Vance", "Watson", "Xu", "Young", "Zimmerman"];

// Category-specific text templates to make queries highly realistic and learnable
const TEMPLATES: Record<QueryCategory, string[]> = {
  [QueryCategory.BILLING]: [
    "Hi, I noticed a double charge of ${AMOUNT} on my credit card for order {ORDER_ID}. Can you please issue a refund?",
    "I was billed twice for my subscription. The invoice shows {ORDER_ID} for {AMOUNT} but I only authorized one charge.",
    "My card on file was charged {AMOUNT} but I cancelled my plan last week. Refund needed ASAP.",
    "Could you send me the invoice for my purchase of {PRODUCT}? Order ID is {ORDER_ID}.",
    "I'm trying to buy {PRODUCT} from {CITY}, but the payment keeps failing with billing code {ERROR_CODE}.",
    "Why was my account charged {AMOUNT}? I didn't purchase anything recently. Order reference is {ORDER_ID}.",
    "I need to update my billing address from {CITY} to my new home. Can you help me?",
    "I am requesting a refund for order {ORDER_ID} since the product arrived damaged. The total amount was ${AMOUNT}."
  ],
  [QueryCategory.TECHNICAL]: [
    "My {DEVICE} is throwing error {ERROR_CODE} when trying to connect to the server. Please help.",
    "The software crashes on startup on my {DEVICE}. It worked fine yesterday but now displays {ERROR_CODE}.",
    "I am unable to login on my {DEVICE}. It keeps spinning and says {ERROR_CODE}.",
    "How do I configure the firmware update on the {PRODUCT}? I'm getting a connection failure.",
    "I'm experiencing slow speeds in {CITY} using {DEVICE}. Is there an outage in my area?",
    "The secure setup for {PRODUCT} fails with code {ERROR_CODE}. I have tried restarting the router already.",
    "Can you help me debug the sync issue on {DEVICE}? It shows database error {ERROR_CODE} constantly.",
    "The application screen is completely frozen on my {DEVICE} after the latest update."
  ],
  [QueryCategory.SHIPPING]: [
    "Where is my package? The tracking for order {ORDER_ID} has not updated in 3 days. It was supposed to ship to {CITY}.",
    "I ordered {PRODUCT} but the delivery address is incorrect. Can I change it to {CITY}?",
    "The delivery status for order {ORDER_ID} says delivered, but I haven't received anything at my home in {CITY}.",
    "How long does shipping to {CITY} take for the {PRODUCT}? I need it before Friday.",
    "My order {ORDER_ID} was shipped via express delivery but it's delayed. Can I get a shipping fee refund?",
    "I received a notification that order {ORDER_ID} was returned to sender. Why did that happen?",
    "Can you provide the tracking number or link for my shipping order {ORDER_ID}?",
    "The package containing my {PRODUCT} is stuck at the local hub in {CITY}. Please check on it."
  ],
  [QueryCategory.PRODUCT]: [
    "Does {PRODUCT} support integration with {DEVICE}? I couldn't find this spec in the manual.",
    "I am looking at {PRODUCT} and wanted to know if it comes with a warranty. What does the warranty cover?",
    "Can you tell me the physical dimensions and weight of {PRODUCT}? I want to make sure it fits my desk.",
    "Is {PRODUCT} compatible with Mac and Windows? I plan to use it on my {DEVICE}.",
    "What is the battery life of {PRODUCT}? Does it support fast charging on {DEVICE}?",
    "How do I clean and maintain my {PRODUCT}? The user guide doesn't explain this well.",
    "Is {PRODUCT} currently in stock in any physical stores near {CITY}?",
    "I want to compare {PRODUCT} with other models. What are the key advantages of this version?"
  ],
  [QueryCategory.ACCOUNT]: [
    "I forgot my password and the reset email is not arriving. My email is {EMAIL}.",
    "How can I delete my profile? I no longer wish to use this service. My customer name is {NAME}.",
    "My profile has been suspended for security reasons, it says error {ERROR_CODE}. How can I unlock it?",
    "I need to change my registered email from {EMAIL} to another address.",
    "Someone tried to access my profile from a different location. Can you verify my login history?",
    "How do I enable two-factor authentication (2FA) on my {DEVICE}? I want to secure my account.",
    "I cannot change my username. It keeps saying invalid format. Please help.",
    "I want to merge my old profile with my new one. Both are under {NAME}."
  ]
};

export interface CustomerRecord {
  customer_id: string;
  customer_name: string;
  email: string;
  phone: string;
  city: string;
  state: string;
  account_created_date: string;
  customer_segment: string;
  support_history_count: number;
}

export interface OrderRecord {
  order_id: string;
  customer_id: string;
  product_name: string;
  product_category: string;
  order_date: string;
  order_amount: number;
  payment_status: string;
  delivery_status: string;
  expected_delivery_date: string;
  refund_status: string;
  warranty_status: string;
}

export interface SupportQueryRecord {
  ticket_id: string;
  customer_id: string;
  order_id: string;
  customer_query: string;
  query_category: string;
  sentiment: string;
  priority: string;
  created_date: string;
  resolution_status: string;
  escalation_status: string;
}

export function generateFullDataset(customerCount: number = 10000): {
  customers: CustomerRecord[];
  orders: OrderRecord[];
  queries: SupportQueryRecord[];
  mappedQueriesForTraining: SupportQuery[];
} {
  const customers: CustomerRecord[] = [];
  const orders: OrderRecord[] = [];
  const queries: SupportQueryRecord[] = [];
  const mappedQueriesForTraining: SupportQuery[] = [];

  console.log(`Generating ${customerCount} customers...`);
  
  // 1. Generate 10,000 Customers
  for (let i = 0; i < customerCount; i++) {
    const custId = `CUST-${10000 + i}`;
    const location = randomChoice(CITIES_WITH_STATES);
    const firstName = randomChoice(FIRST_NAMES);
    const lastName = randomChoice(LAST_NAMES);
    const customerName = `${firstName} ${lastName}`;
    const email = `${firstName.toLowerCase()}.${lastName.toLowerCase()}@example.com`;
    const phone = `+1-555-${randomRange(100, 999).toString().padStart(3, "0")}-${randomRange(1000, 9999).toString().padStart(4, "0")}`;
    
    // Account creation date over last 3 years
    const acctDate = new Date();
    acctDate.setDate(acctDate.getDate() - randomRange(30, 1000));
    const accountCreatedDate = acctDate.toISOString().split("T")[0];
    
    const customerSegment = randomChoice(["Standard", "Premium", "VIP"]);
    const supportHistoryCount = randomRange(0, 5);

    customers.push({
      customer_id: custId,
      customer_name: customerName,
      email,
      phone,
      city: location.city,
      state: location.state,
      account_created_date: accountCreatedDate,
      customer_segment: customerSegment,
      support_history_count: supportHistoryCount
    });

    // 2. Generate 1 to 5 Orders for each Customer
    const numOrders = randomRange(1, 5);
    const customerOrders: OrderRecord[] = [];
    
    for (let j = 0; j < numOrders; j++) {
      const orderId = `ORD-${randomRange(100000, 999999)}`;
      const product = randomChoice(PRODUCTS);
      
      // Order date is after account created
      const oDate = new Date(accountCreatedDate);
      oDate.setDate(oDate.getDate() + randomRange(1, 20));
      const orderDate = oDate.toISOString().split("T")[0];
      
      const paymentStatus = randomChoice(["Paid", "Paid", "Paid", "Paid", "Paid", "Unpaid", "Refunded", "Processing Refund"]);
      const deliveryStatus = randomChoice(["Delivered", "Delivered", "Delivered", "In Transit", "Not Shipped", "Cancelled"]);
      
      const expDate = new Date(orderDate);
      expDate.setDate(expDate.getDate() + randomRange(3, 7));
      const expectedDeliveryDate = expDate.toISOString().split("T")[0];
      
      let refundStatus = "None";
      if (paymentStatus === "Refunded") refundStatus = "Completed";
      else if (paymentStatus === "Processing Refund") refundStatus = "Processing";

      let warrantyStatus = "No Warranty";
      if (product.category !== "Accessories") {
        const orderAgeDays = (Date.now() - new Date(orderDate).getTime()) / (1000 * 60 * 60 * 24);
        warrantyStatus = orderAgeDays > 365 ? "Expired" : "Active";
      }

      const oRec: OrderRecord = {
        order_id: orderId,
        customer_id: custId,
        product_name: product.name,
        product_category: product.category,
        order_date: orderDate,
        order_amount: product.price,
        payment_status: paymentStatus,
        delivery_status: deliveryStatus,
        expected_delivery_date: expectedDeliveryDate,
        refund_status: refundStatus,
        warranty_status: warrantyStatus
      };
      
      orders.push(oRec);
      customerOrders.push(oRec);
    }
  }

  console.log(`Generated ${orders.length} orders. Generating 10,000 support queries...`);

  // 3. Generate 10,000 Support Queries linked to random customers & their orders
  const categories = Object.values(QueryCategory);
  const sentiments = Object.values(Sentiment);
  const priorities = ["low", "medium", "high", "urgent"];
  const resolutionStatuses = ["new", "pending", "in_progress", "resolved", "escalated"];

  for (let i = 0; i < 10000; i++) {
    const tktId = `TKT-${100000 + i}`;
    
    // Select a random customer
    const customer = randomChoice(customers);
    
    // Get that customer's orders
    const custOrders = orders.filter(o => o.customer_id === customer.customer_id);
    const order = randomChoice(custOrders) || { order_id: "ORD-UNKNOWN", product_name: "Pro Headphones X", order_amount: 299 };
    
    // Choose a category
    const category = randomChoice(categories);
    const sentiment = randomChoice(sentiments);
    const priority = randomChoice(priorities);
    
    // Created date over the last 30 days
    const cDate = new Date();
    cDate.setDate(cDate.getDate() - randomRange(0, 30));
    const createdDate = cDate.toISOString().split("T")[0];
    
    const resolutionStatus = randomChoice(resolutionStatuses);
    const escalationStatus = resolutionStatus === "escalated" ? "Yes" : "No";

    const device = randomChoice(DEVICES);
    const errorCode = randomChoice(ERROR_CODES);

    // Grab templates for this category
    const templates = TEMPLATES[category];
    let queryText = randomChoice(templates);

    // Substitute placeholders
    queryText = queryText
      .replace(/{ORDER_ID}/g, order.order_id)
      .replace(/{CITY}/g, customer.city)
      .replace(/{PRODUCT}/g, order.product_name)
      .replace(/{AMOUNT}/g, order.order_amount.toString())
      .replace(/{DEVICE}/g, device)
      .replace(/{ERROR_CODE}/g, errorCode)
      .replace(/{NAME}/g, customer.customer_name)
      .replace(/{EMAIL}/g, customer.email);

    if (Math.random() < 0.1) {
      queryText = queryText.toLowerCase();
    }
    if (Math.random() < 0.05) {
      queryText = queryText + " please assist asap.";
    }

    queries.push({
      ticket_id: tktId,
      customer_id: customer.customer_id,
      order_id: order.order_id,
      customer_query: queryText,
      query_category: category,
      sentiment,
      priority,
      created_date: createdDate,
      resolution_status: resolutionStatus,
      escalation_status: escalationStatus
    });

    // Also map to traditional SupportQuery interface for ML / DL training
    mappedQueriesForTraining.push({
      id: tktId,
      query: queryText,
      category,
      orderId: order.order_id,
      city: customer.city,
      product: order.product_name,
      amount: order.order_amount,
      device,
      errorCode,
      customerName: customer.customer_name,
      sentiment,
      timestamp: new Date(createdDate).toISOString()
    });
  }

  return {
    customers,
    orders,
    queries,
    mappedQueriesForTraining
  };
}

export function saveFullDatasetToCSVAndJSON(
  data: {
    customers: CustomerRecord[];
    orders: OrderRecord[];
    queries: SupportQueryRecord[];
    mappedQueriesForTraining: SupportQuery[];
  },
  dirPath: string
) {
  // Save to the specified customer_support_ai directory
  if (!fs.existsSync(dirPath)) {
    fs.mkdirSync(dirPath, { recursive: true });
  }

  // Also save in root 'data/' directory as requested by the user
  const rootDataDir = path.join(process.cwd(), "data");
  if (!fs.existsSync(rootDataDir)) {
    fs.mkdirSync(rootDataDir, { recursive: true });
  }

  const escapeCSV = (str: string) => {
    if (typeof str !== "string") return str;
    const escaped = str.replace(/"/g, '""');
    return `"${escaped}"`;
  };

  // 1. Write customers.csv
  const customerHeaders = ["Customer ID", "Customer Name", "Email", "Phone Number", "City", "State", "Account Created Date", "Customer Segment", "Support History Count"];
  const customerRows = [customerHeaders.join(",")];
  for (const c of data.customers) {
    customerRows.push([
      c.customer_id,
      escapeCSV(c.customer_name),
      escapeCSV(c.email),
      escapeCSV(c.phone),
      escapeCSV(c.city),
      escapeCSV(c.state),
      c.account_created_date,
      c.customer_segment,
      c.support_history_count
    ].join(","));
  }

  fs.writeFileSync(path.join(dirPath, "customers.csv"), customerRows.join("\n"), "utf-8");
  fs.writeFileSync(path.join(rootDataDir, "customers.csv"), customerRows.join("\n"), "utf-8");

  // 2. Write orders.csv
  const orderHeaders = ["Order ID", "Customer ID", "Product Name", "Product Category", "Order Date", "Order Amount", "Payment Status", "Delivery Status", "Expected Delivery Date", "Refund Status", "Warranty Status"];
  const orderRows = [orderHeaders.join(",")];
  for (const o of data.orders) {
    orderRows.push([
      o.order_id,
      o.customer_id,
      escapeCSV(o.product_name),
      escapeCSV(o.product_category),
      o.order_date,
      o.order_amount,
      o.payment_status,
      o.delivery_status,
      o.expected_delivery_date,
      o.refund_status,
      o.warranty_status
    ].join(","));
  }

  fs.writeFileSync(path.join(dirPath, "orders.csv"), orderRows.join("\n"), "utf-8");
  fs.writeFileSync(path.join(rootDataDir, "orders.csv"), orderRows.join("\n"), "utf-8");

  // 3. Write support_queries.csv
  const queryHeaders = ["Ticket ID", "Customer ID", "Order ID", "Customer Query", "Query Category", "Sentiment", "Priority", "Created Date", "Resolution Status", "Escalation Status"];
  const queryRows = [queryHeaders.join(",")];
  for (const q of data.queries) {
    queryRows.push([
      q.ticket_id,
      q.customer_id,
      q.order_id,
      escapeCSV(q.customer_query),
      q.query_category,
      q.sentiment,
      q.priority,
      q.created_date,
      q.resolution_status,
      q.escalation_status
    ].join(","));
  }

  fs.writeFileSync(path.join(dirPath, "support_queries.csv"), queryRows.join("\n"), "utf-8");
  fs.writeFileSync(path.join(rootDataDir, "support_queries.csv"), queryRows.join("\n"), "utf-8");

  // 4. Also save original formatted customer_support_queries JSON and CSV for model training pipeline compatibility
  fs.writeFileSync(path.join(dirPath, "customer_support_queries.json"), JSON.stringify(data.mappedQueriesForTraining, null, 2), "utf-8");
  
  const originalHeaders = [
    "id", "query", "category", "orderId", "city", "product", "amount", "device", "errorCode", "customerName", "sentiment", "timestamp"
  ];
  const originalRows = [originalHeaders.join(",")];
  for (const q of data.mappedQueriesForTraining) {
    originalRows.push([
      q.id,
      escapeCSV(q.query),
      q.category,
      q.orderId,
      escapeCSV(q.city),
      escapeCSV(q.product),
      q.amount,
      escapeCSV(q.device),
      q.errorCode,
      escapeCSV(q.customerName),
      q.sentiment,
      q.timestamp
    ].join(","));
  }
  fs.writeFileSync(path.join(dirPath, "customer_support_queries.csv"), originalRows.join("\n"), "utf-8");

  console.log(`Successfully generated and written datasets!`);
  console.log(`- 10,000 Customers written to data/customers.csv`);
  console.log(`- ${data.orders.length} Orders written to data/orders.csv`);
  console.log(`- 10,000 Support Queries written to data/support_queries.csv`);
  console.log(`- ML Pipeline compatibility files generated.`);
}

// Deprecated original generator mapped to new generator to prevent import breaks
export function generateSyntheticDataset(count: number = 10000): SupportQuery[] {
  const result = generateFullDataset(count);
  return result.mappedQueriesForTraining;
}

export function saveDatasetToCSVAndJSON(queries: SupportQuery[], dirPath: string) {
  // Just in case, map SupportQuery back to full datasets to fulfill start conditions
  const fullData = generateFullDataset(queries.length);
  saveFullDatasetToCSVAndJSON(fullData, dirPath);
}

File: app/applet/customer_support_ai/data_generator.py
# -*- coding: utf-8 -*-
"""
High-Performance programmatically synthetic dataset generator for Customer Support AI.
Generates 10,000 highly structured, cohesive customer profiles, matching orders, and tickets.
"""

import os
import csv
import json
import random
from datetime import datetime, timedelta

from utils import DATA_DIR, WORKSPACE_DATA_DIR

# --- MOCK DATA POOLS ---
FIRST_NAMES = ["James", "Mary", "John", "Patricia", "Robert", "Jennifer", "Michael", "Linda", "William", "Elizabeth",
               "David", "Barbara", "Richard", "Susan", "Joseph", "Jessica", "Thomas", "Sarah", "Charles", "Karen",
               "Christopher", "Nancy", "Daniel", "Lisa", "Matthew", "Betty", "Anthony", "Margaret", "Mark", "Sandra"]

LAST_NAMES = ["Smith", "Johnson", "Williams", "Brown", "Jones", "Miller", "Davis", "Garcia", "Rodriguez", "Wilson",
              "Martinez", "Anderson", "Taylor", "Thomas", "Hernandez", "Moore", "Martin", "Jackson", "Thompson", "White",
              "Lopez", "Lee", "Gonzalez", "Harris", "Clark", "Lewis", "Robinson", "Walker", "Perez", "Hall"]

CITIES = [
    ("Austin", "TX"), ("San Francisco", "CA"), ("New York", "NY"), ("Seattle", "WA"),
    ("Denver", "CO"), ("Miami", "FL"), ("Chicago", "IL"), ("Boston", "MA"),
    ("Portland", "OR"), ("Los Angeles", "CA"), ("Atlanta", "GA"), ("Phoenix", "AZ"),
    ("Dallas", "TX"), ("San Diego", "CA"), ("Salt Lake City", "UT"), ("Nashville", "TN")
]

SEGMENTS = ["Retail Customer", "Premium VIP", "Enterprise Corporate", "SME Business"]

PRODUCTS = [
    ("Secure Router Pro", "Hardware", 199.99),
    ("Mesh Extender Dual-Band", "Hardware", 89.99),
    ("Gigabit Switch 8-Port", "Hardware", 49.99),
    ("SaaS Cloud Firewall Plus", "Subscription", 29.99),
    ("Enterprise VPN Client license", "Subscription", 14.99),
    ("Premium Support SLA Ticket", "Subscription", 9.99)
]

QUERY_TEMPLATES = [
    # Billing templates
    ("I noticed a double charge of {amount} on my credit card for {product}. Please issue a refund.", "Billing", "Negative", "High", "No"),
    ("Can you please send me the tax invoice for order {order_id}? I need it for my expense claim.", "Billing", "Neutral", "Medium", "No"),
    ("Why was my credit card charged {amount} without authorization? I want a refund immediately.", "Billing", "Negative", "High", "No"),
    ("My billing subscription renewal failed with error {error_code}. Can you assist with the payment?", "Billing", "Negative", "Medium", "No"),
    ("I was overcharged for {product}. The catalog says it is on discount but I paid full price.", "Billing", "Negative", "Medium", "No"),
    
    # Technical templates
    ("My {product} is crashing constantly. The system log shows error code {error_code}.", "Technical", "Negative", "High", "Yes"),
    ("Can you provide instructions on how to install the firmware updates for my secure router pro?", "Technical", "Neutral", "Medium", "No"),
    ("I am getting a timeout exception ERR_DB_TIMEOUT_504 when attempting to login to my account profile.", "Technical", "Negative", "High", "Yes"),
    ("The mobile application is freezing on the splash screen after the latest software update.", "Technical", "Negative", "Medium", "No"),
    ("I need help with port forwarding settings on my dual-band mesh extender.", "Technical", "Neutral", "Medium", "No"),
    
    # Shipping templates
    ("My order {order_id} has not arrived yet. It has been in transit for two weeks. Where is it?", "Shipping", "Negative", "High", "No"),
    ("Can you update the shipping address for order {order_id} to 120 Main Street, Austin TX?", "Shipping", "Neutral", "Medium", "No"),
    ("The tracking link for my parcel is broken and shows no updates from the logistics courier.", "Shipping", "Negative", "Medium", "No"),
    ("My package arrived yesterday but the retail box was damaged and the secure router pro is cracked.", "Shipping", "Negative", "High", "Yes"),
    ("When will my order ship? The expected delivery date was supposed to be yesterday.", "Shipping", "Negative", "Medium", "No"),
    
    # Product templates
    ("Are the specs of secure router pro dual-band or tri-band? I need to verify specifications.", "Product", "Neutral", "Low", "No"),
    ("Is the premium support SLA license compatible with older gigabit switches?", "Product", "Neutral", "Low", "No"),
    ("What are the exact dimensions and weight of the mesh extender hardware?", "Product", "Neutral", "Low", "No"),
    ("Does the Secure Router Pro come with a standard replacement warranty?", "Product", "Neutral", "Low", "No"),
    ("Can I use the Enterprise VPN client on multiple devices simultaneously?", "Product", "Neutral", "Medium", "No")
]

ERROR_CODES = ["ERR_DB_TIMEOUT_504", "ERR_AUTH_FAILED_401", "ERR_CONN_RESET_104", "ERR_SYS_CRASH_500"]

def generate_full_dataset(num_customers: int = 10000) -> dict:
    random.seed(42) # Deterministic generation
    
    customers = []
    orders = []
    queries = []

    start_date = datetime(2025, 1, 1)

    for i in range(1, num_customers + 1):
        # 1. Generate Customer Profile
        cust_id = f"CUST-{i:06d}"
        first = random.choice(FIRST_NAMES)
        last = random.choice(LAST_NAMES)
        name = f"{first} {last}"
        email = f"{first.lower()}.{last.lower()}{random.randint(10, 99)}@example.com"
        phone = f"+1-{random.randint(200, 999)}-{random.randint(200, 999)}-{random.randint(1000, 9999)}"
        city, state = random.choice(CITIES)
        created_days_ago = random.randint(10, 365)
        created_date = start_date + timedelta(days=created_days_ago)
        segment = random.choice(SEGMENTS)
        support_count = random.randint(0, 5)

        customers.append({
            "Customer ID": cust_id,
            "Customer Name": name,
            "Email": email,
            "Phone Number": phone,
            "City": city,
            "State": state,
            "Account Created Date": created_date.strftime("%Y-%m-%d"),
            "Customer Segment": segment,
            "Support History Count": support_count
        })

        # 2. Generate Orders for Customer (1 to 5 orders)
        num_orders = random.randint(1, 5)
        cust_orders = []
        for j in range(1, num_orders + 1):
            order_id = f"ORD-{i:06d}-{j}"
            prod_name, prod_cat, prod_price = random.choice(PRODUCTS)
            
            # Order Date must be after account creation
            order_days_after = random.randint(1, 30)
            order_date = created_date + timedelta(days=order_days_after)
            
            # Logistics status
            pay_status = random.choices(["Paid", "Pending", "Failed"], weights=[0.85, 0.10, 0.05])[0]
            del_status = "Unshipped"
            if pay_status == "Paid":
                del_status = random.choices(["Delivered", "In Transit", "Unshipped"], weights=[0.70, 0.20, 0.10])[0]

            expected_date = order_date + timedelta(days=5)
            
            refund_status = "N/A"
            if pay_status == "Paid" and random.random() < 0.08:
                refund_status = random.choice(["Requested", "Refunded"])

            warranty_status = "Active" if prod_cat == "Hardware" and random.random() < 0.90 else "N/A"

            order_record = {
                "Order ID": order_id,
                "Customer ID": cust_id,
                "Product Name": prod_name,
                "Product Category": prod_cat,
                "Order Date": order_date.strftime("%Y-%m-%d"),
                "Order Amount": prod_price,
                "Payment Status": pay_status,
                "Delivery Status": del_status,
                "Expected Delivery Date": expected_date.strftime("%Y-%m-%d"),
                "Refund Status": refund_status,
                "Warranty Status": warranty_status
            }
            orders.append(order_record)
            cust_orders.append(order_record)

        # 3. Generate Support Query for Customer (every customer has exactly 1 linked ticket in this dataset)
        ticket_id = f"TKT-{i:06d}"
        linked_order = random.choice(cust_orders)
        
        # Select query template
        tpl_text, q_cat, q_sent, q_prio, q_esc = random.choice(QUERY_TEMPLATES)
        
        # Populate template parameters
        err = random.choice(ERROR_CODES)
        amt = f"${linked_order['Order Amount']:.2f}"
        prod = linked_order['Product Name']
        ord_ref = linked_order['Order ID']

        query_text = tpl_text.format(amount=amt, product=prod, order_id=ord_ref, error_code=err)

        # Resolution Status mapping
        res_status = random.choices(["resolved", "pending", "escalated"], weights=[0.60, 0.30, 0.10])[0]
        if q_esc == "Yes":
            res_status = "escalated"

        # Ticket date must be after order date
        order_dt = datetime.strptime(linked_order["Order Date"], "%Y-%m-%d")
        ticket_days_after = random.randint(1, 15)
        ticket_date = order_dt + timedelta(days=ticket_days_after)

        queries.append({
            "Ticket ID": ticket_id,
            "Customer ID": cust_id,
            "Order ID": ord_ref,
            "Customer Query": query_text,
            "Query Category": q_cat,
            "Sentiment": q_sent,
            "Priority": q_prio,
            "Created Date": ticket_date.strftime("%Y-%m-%d"),
            "Resolution Status": res_status,
            "Escalation Status": "Yes" if res_status == "escalated" else "No"
        })

    return {
        "customers": customers,
        "orders": orders,
        "queries": queries
    }

def save_full_dataset_to_csv_and_json(dataset: dict, target_dir: str):
    # Ensure folder is created
    os.makedirs(target_dir, exist_ok=True)
    os.makedirs(WORKSPACE_DATA_DIR, exist_ok=True)

    # 1. Save Customers CSV
    for folder in [target_dir, WORKSPACE_DATA_DIR]:
        cust_path = os.path.join(folder, "customers.csv")
        with open(cust_path, "w", newline="", encoding="utf-8") as f:
            writer = csv.DictWriter(f, fieldnames=dataset["customers"][0].keys())
            writer.writeheader()
            writer.writerows(dataset["customers"])

        # 2. Save Orders CSV
        order_path = os.path.join(folder, "orders.csv")
        with open(order_path, "w", newline="", encoding="utf-8") as f:
            writer = csv.DictWriter(f, fieldnames=dataset["orders"][0].keys())
            writer.writeheader()
            writer.writerows(dataset["orders"])

        # 3. Save Queries CSV
        query_path = os.path.join(folder, "support_queries.csv")
        with open(query_path, "w", newline="", encoding="utf-8") as f:
            writer = csv.DictWriter(f, fieldnames=dataset["queries"][0].keys())
            writer.writeheader()
            writer.writerows(dataset["queries"])

        # 4. Save Support Queries JSON (for easy training parsing)
        json_path = os.path.join(folder, "customer_support_queries.json")
        with open(json_path, "w", encoding="utf-8") as f:
            # Map values to match exactly what ml_module.ts expects
            json_list = []
            for q in dataset["queries"]:
                cust_match = [c for c in dataset["customers"] if c["Customer ID"] == q["Customer ID"]][0]
                json_list.append({
                    "id": q["Ticket ID"],
                    "query": q["Customer Query"],
                    "category": q["Query Category"],
                    "sentiment": q["Sentiment"],
                    "priority": q["Priority"],
                    "customerName": cust_match["Customer Name"],
                    "orderId": q["Order ID"]
                })
            json.dump(json_list, f, indent=2)

    print(f"Programmatic dataset files successfully saved to: {target_dir} and {WORKSPACE_DATA_DIR}")

if __name__ == "__main__":
    ds = generate_full_dataset(10000)
    save_full_dataset_to_csv_and_json(ds, DATA_DIR)
